Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift

```diff
@@ -0,0 +1,154 @@
+  func audioVideoPlayer(didPausePlayback player: AudioVideoPlayer) {}
+  func audioVideoPlayer(didBeginReplay player: AudioVideoPlayer) {}
+  func audioVideoPlayer(didBeginBuffering player: AudioVideoPlayer) {}
+  func audioVideoPlayer(didEndBuffering player: AudioVideoPlayer) {}
+  func audioVideoPlayer(_ player: AudioVideoPlayer, didProgressToTime seconds: Double) {}
+}
+
+public class AudioVideoPlayer: AVPlayer {
+  public static let defaultTimescale = CMTimeScale(kCMTimeMaxTimescale)
+
+  public private(set) lazy var playbackInterval: (startAt: Double, endAt: Double) = (0, duration)
+  public private(set) var isPlaying = false
+  public var allowLooping = true
+  public weak var delegate: AudioVideoPlayerDelegate?
+  
+  private var timeObserver: Any?
+
+  public override init() {
+    super.init()
+    setupTimeObserver()
+  }
+
+  public override init(url: URL) {
+    super.init(url: url)
+    setupTimeObserver()
+  }
+
+  override init(playerItem item: AVPlayerItem?) {
+    super.init(playerItem: item)
+    setupTimeObserver()
+  }
+  
+  convenience public init?(urlString: String) {
+    guard let url = URL(string: urlString) else { return nil }
+    self.init(url: url)
+  }
+}
+
+// MARK: - API
+public extension AudioVideoPlayer {
+  override func play() {
+    super.play()
+    isPlaying = true
+    delegate?.audioVideoPlayer(didBeginPlayback: self)
+  }
+
+  override func pause() {
+    super.pause()
+    isPlaying = false
+    delegate?.audioVideoPlayer(didPausePlayback: self)
+  }
+
+  override func replaceCurrentItem(with item: AVPlayerItem?) {
+    super.replaceCurrentItem(with: item)
+    playbackInterval = (0, duration)
+    setupTimeObserver()
+  }
+
+  /// Set `playbackInterval` and force player to start playing from the `startAt` timestamp.
+  func setPlaybackInterval(
+    startAt: Double,
+    endAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
+    guard startAt < endAt else { fatalError("`startAt` should be less than `endAt`") }
+    setPlaybackPosition(startAt: startAt)
+    playbackInterval = (startAt, endAt)
+  }
+
+  /// Update `playbackInterval` and keep playing from the current location.
+  func updatePlaybackInterval(startAt: Double, endAt: Double) {
+    guard startAt < endAt else { fatalError("`startAt` should be less than `endAt`") }
+    playbackInterval = (startAt, endAt)
+    setupTimeObserver()
+  }
+
+  func setPlaybackPosition(
+    startAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
+    removeTimeObserver()
+    seek(to: CMTime(seconds: startAt, preferredTimescale: timescale)) { _ in
+      self.setupTimeObserver()
+    }
+  }
+
+  var duration: Double {
+    currentItem?.asset.duration.seconds ?? 0
+  }
+}
+
+// MARK: - Private Methods
+private extension AudioVideoPlayer {
+  func setupTimeObserver() {
+    guard timeObserver == nil else { return }
+
+    let timeInterval = CMTimeMake(value: 1, timescale: 10)
```

Comment:
So the observer will trigger every 1/10th of a second. What if we need it to be triggered more/less often? Should we make this configurable?

Reply:
We should, of course.